sale_loyalty_incompatibility/tests/test_sale_coupon_incompatibility.py

```diff
@@ -1,99 +1,82 @@
-        return coupon_program_form.save()
-
-    def _generate_coupons(self, program, number=1):
-        """Helper method to easily generate coupons in the test cases"""
-        self.env["coupon.generate.wizard"].with_context(active_id=program.id).create(
-            {"generation_type": "nbr_coupon", "nbr_coupons": number}
-        ).generate_coupon()
-        return program.coupon_ids
-
-    def _create_sale(self, partner, products=False):
-        """Helper method to create sales in the test cases"""
-        sale_form = Form(self.env["sale.order"])
-        sale_form.partner_id = partner
-        if not products:
-            products = self.product_a
-        for product in products:
-            with sale_form.order_line.new() as line_form:
-                line_form.product_id = product
-                line_form.product_uom_qty = 3
-        return sale_form.save()
-
-    def _apply_coupon(self, order, code, bp=False):
-        """Helper method to apply either coupon or progam codes. It ensures that the
-        UserError exception is raised as well."""
-        self.env["sale.coupon.apply.code"].with_context(
-            active_id=order.id, bp=bp
-        ).create({"coupon_code": code}).process_coupon()
-
+class TestLoyaltyIncompatibility(TestSaleCouponCommon):
     def test_01_program_incompatibilities(self):
         """A program with no code and customer application limit won't be applied
         once the limit is reached"""
         # First we test the regular behavior
-        promotion_1 = self._create_coupon_program(self.product_a, code="TEST-00000002")
-        promotion_2 = self._create_coupon_program(self.product_b, code="TEST-00000003")
-        coupons_2 = (x for x in self._generate_coupons(promotion_2, 2))
-        promotion_3 = self._create_coupon_program(self.product_c, code="TEST-00000004")
-        coupons_3 = (x for x in self._generate_coupons(promotion_3, 2))
-        sale_1 = self._create_sale(
-            self.partner_1, (self.product_a + self.product_b + self.product_c)
+        self.immediate_promotion_program.active = False
+        self.code_promotion_program.reward_ids.reward_type = "discount"
+        self.code_promotion_program.reward_ids.discount = 10
+        promotion_1 = self.code_promotion_program
+        promotion_2 = self.code_promotion_program_with_discount
+        self.env["loyalty.generate.wizard"].with_context(active_id=promotion_1).create(
+            {
+                "mode": "selected",
+                "customer_ids": self.steve,
+                "points_granted": 1,
+            }
+        ).generate_coupons()
+        coupon_promotion_1 = promotion_1.coupon_ids
+        order = self.empty_order
+        order.write(
+            {
+                "order_line": [
+                    (
+                        0,
+                        False,
+                        {
+                            "product_id": self.product_A.id,
```

Comment:
For the sake of clarity I'd suggest you to create sale lines using `Form`

Reply:
i created with write because into original test from odoo they used the create method

Reply:
Yeah, I know, but don't lose the chance to make it better (more compact and expressive)

Reply:
Done @chienandalu, please review